server/src/main/java/io/druid/server/coordination/BatchDataSegmentAnnouncer.java

```diff
@@ -87,36 +86,41 @@ public void announceSegment(DataSegment segment) throws IOException
     }
 
     synchronized (lock) {
-      // create new batch
-      if (availableZNodes.isEmpty()) {
-        SegmentZNode availableZNode = new SegmentZNode(makeServedSegmentPath());
-        availableZNode.addSegment(segment);
-
-      log.info("Announcing segment[%s] at path[%s]", segment.getIdentifier(), availableZNode.getPath());
-      announcer.announce(availableZNode.getPath(), availableZNode.getBytes());
-      segmentLookup.put(segment, availableZNode);
-      availableZNodes.add(availableZNode);
-    } else { // update existing batch
-      Iterator<SegmentZNode> iter = availableZNodes.iterator();
-      boolean done = false;
-      while (iter.hasNext() && !done) {
-        SegmentZNode availableZNode = iter.next();
+      // try to update existing batch
+      for (SegmentZNode availableZNode : availableZNodes) {
         if (availableZNode.getBytes().length + newBytesLen < config.getMaxBytesPerNode()) {
-          availableZNode.addSegment(segment);
-
-            log.info("Announcing segment[%s] at path[%s]", segment.getIdentifier(), availableZNode.getPath());
-            announcer.update(availableZNode.getPath(), availableZNode.getBytes());
-            segmentLookup.put(segment, availableZNode);
-
-            if (availableZNode.getCount() >= config.getSegmentsPerNode()) {
-              availableZNodes.remove(availableZNode);
-            }
-
-            done = true;
+          SegmentZNode announced = announceSegmentTo(segment, availableZNode);
+          if (announced.getCount() >= config.getSegmentsPerNode()) {
+            availableZNodes.remove(availableZNode);
           }
+          return;
         }
       }
+      // create new batch
+      SegmentZNode announced = announceSegmentTo(segment, null);
```

Comment:
It will probably be cleaner to create a new SegmentZNode and pass it, instead of the announceSegmentTo creating a new node